pkg/urbit/include/c/defs.h

```diff
@@ -157,4 +159,108 @@
 #     define c3_fopen(a, b) ({                                  \
         fopen(a, b);})
 
+      /// Reads the contents of a file into memory.
+      ///
+      /// This function correctly handles the case in which read() reads fewer
+      /// bytes than was requested by retrying until all bytes have been read or
+      /// until an error occurs, whichever comes first.
+      ///
+      /// If interrupted by a signal or if the read() call would block, this
+      /// function will retry up to a specified maximum number of attempts.
+      ///
+      /// @param[in]  fd_i        Open file descriptor to read from.
+      /// @param[out] data_v      Destination buffer.
+      /// @param[in]  data_len_i  Length of `data_v` in bytes.
+      ///
+      /// @return 1  Complete read succeeded.
+      /// @return 0  Otherwise.
+      static inline c3_t
+      c3_read(c3_i fd_i, void* const data_v, const size_t data_len_i)
+      {
+        static const size_t max_attempts_i    = 100;
+        size_t              attempts_i        = 0;
+        c3_y*               ptr_y             = data_v;
+        ssize_t             bytes_remaining_i = data_len_i;
+        do {
+          attempts_i++;
+          ssize_t bytes_read_i = read(fd_i, ptr_y, bytes_remaining_i);
+          if ( -1 == bytes_read_i ) {
+            if ( (EINTR == errno || EAGAIN == errno || EWOULDBLOCK == errno)
+                 && attempts_i <= max_attempts_i )
+            {
+              continue;
+            }
+            else {
+              fprintf(stderr,
+                      "read: failed after %lu attempts: %s\r\n",
+                      attempts_i,
+                      strerror(errno));
+              return 0;
+            }
+          }
+          else if ( 0 == bytes_read_i ) {
+            break;
+          }
```

Comment:
Needing to handle `EOF` here makes me concerned that this signature is insufficient. Reads may correctly produce less than the requested number of bytes, but this function swallows that measurement, which could lead to bogus data being propagated in the target buffer. I'm not sure what the best solution is. The function could return `-errno` on error, `0` on a successful read of the amount requested, or some positive integer on a successful but truncated (due to `EOF`) read (with the invariant that otherwise partial reads have already been retried). Or it could just match the `read()` signature.

Also, almost all calls to this function, already print on error. It seems like this function should either propagate/preserve `errno` for those call sites, or additional take a caption string as a parameter and interpolate that into the error printf.

Reply:
One other point about this signature, from the man page:

>  On error, -1 is returned, and errno is set appropriately. In this case it is left unspecified whether the file position (if any) changes.

That's pretty shocking to me. Combining it with the performance argument in #5889 makes me wonder if we should just always use `pread()` in cases where offsets apply. Otherwise, seems like we would logically need an `lseek()` in this function when we retry on error and the fd has an offset, and I don't know how we would distinguish that case. What do you think?

Reply:
> Or it could just match the read() signature.

This seems like the cleanest solution. So `-1` will indicate failure (and `errno` will indicate the specific failure), `0` will indicate that `fd_i` was already at EOF, and `>0` will indicate the number of bytes read into `data_v` (which is guaranteed to be in the range `(0, data_len_i]`.

Reply:
> On error, -1 is returned, and errno is set appropriately. In this case it is left unspecified whether the file position (if any) changes.

Wow, this is terrible.

Reply:
I updated `c3_read()` to track the current offset and reset it to its pre-`read()` fail value when a call to `read()` failed. The only added cost for a call to `c3_read()` for which no `read()` calls fail is a single `lseek()` at the beginning used to compute the current offset (which itself should be relatively cheap).

Reply:
That looks like it would work correctly for files that we read entirely, but it significantly limits the applicability of this wrapper function. At the very least, we now have a bunch of redundant `lseek()` calls in `events.c`. And a bunch of refactored call sites in this PR will be wrong, see `ESPIPE` in the `lseek()` man-page.

Reply:
After speaking live, we've decided to require that `c3_read()` be called with a seekable file.